W5500.device.nut

```diff
@@ -0,0 +1,2256 @@
+                bytes = 0;
+            }
+
+            total += mem_size;
+            _socketMemory[dir][socket] = bytes;
+            local bsb = _getSocketRegBlockSelectBit(socket);
+            writeReg(addr, bsb, mem_size);
+        }
+        return this;
+    }
+
+
+    // ***************************************************************************
+    // getMemory
+    // Returns: the buffer size
+    // Parameters:
+    //      dir - a string containing the transmition direction, accepted values
+    //            are "tx" or "rx"
+    // ***************************************************************************
+    function getMemory(dir, socket) {
+        return _socketMemory[dir][socket];
+    }
+
+    // ***************************************************************************
+    // setMode
+    // Returns: this
+    // Parameters:
+    //      mode - select mode using MODE constants or-ed together
+    // ***************************************************************************
+    function setMode(mode) {
+        writeReg(W5500_MODE, W5500_COMMON_REGISTER, mode);
+        return this;
+    }
+
+    // ***************************************************************************
+    // setSocketMode
+    // Returns: this
+    // Parameters:
+    //      socket - select the socket using an integer 0-3
+    //      mode - select mode using W5500_SOCKET_MODE constant
+    // ***************************************************************************
+    function setSocketMode(socket, mode) {
+        local bsb = _getSocketRegBlockSelectBit(socket)
+        writeReg(W5500_SOCKET_MODE, bsb, mode);
+        return this;
+    }
+
+    // ***************************************************************************
+    // setGatewayAddr
+    // Returns: this
+    // Parameters:
+    //      addr - an array of four integers with the gateway IP address
+    // ***************************************************************************
+    function setGatewayAddr(addr) {
+        local addr = _addrToIP(addr);
+        writeReg(W5500_GATEWAY_ADDR_0, W5500_COMMON_REGISTER, addr[0]);
+        writeReg(W5500_GATEWAY_ADDR_1, W5500_COMMON_REGISTER, addr[1]);
+        writeReg(W5500_GATEWAY_ADDR_2, W5500_COMMON_REGISTER, addr[2]);
+        writeReg(W5500_GATEWAY_ADDR_3, W5500_COMMON_REGISTER, addr[3]);
+        return this;
+    }
+
+    // ***************************************************************************
+    // getGatewayAddr
+    // Returns:  an array of four integers with the gateway IP address
+    // Parameters:
+    //        none
+    // **************************************************************************
+    function getGatewayAddr() {
+        local addr = array(4);
+        // server.log((hardware.millis() - started) + ": DBG setGatewayAddr: " + pformat(addr));
+        addr[0] = readReg(W5500_GATEWAY_ADDR_0, W5500_COMMON_REGISTER);
+        addr[0] = readReg(W5500_GATEWAY_ADDR_1, W5500_COMMON_REGISTER);
```

Comment:
addr[0] is being overwritten?  Should the addr array index increase with each register read?